feat(page_service): timeout-based batching of requests

VladLazar opened the pull request:
## Problem

We don't take advantage of queue depth generated by the compute
on the pageserver. We can process getpage requests more efficiently
by batching them. 

## Summary of changes

Batch up incoming getpage requests that arrive within a configurable time window (`server_side_batch_timeout`).
Then process the entire batch via one `get_vectored` timeline operation.
By default, no merging takes place.

## Testing

* **Functional**: https://github.com/neondatabase/neon/pull/9792
* **Performance**: will be done in staging/pre-prod

problame commented on pageserver/src/pgdatadir_mapping.rs:
```diff
@@ -191,26 +196,184 @@ impl Timeline {
-                nblocks
-            );
-            return Ok(ZERO_PAGE.clone());
+    /// Like [`get_rel_page_at_lsn`], but returns a batch of pages.
+    pub(crate) async fn get_rel_page_at_lsn_batched(
```
```suggestion
    /// Like [`get_rel_page_at_lsn`], but returns a batch of pages.
    ///
    /// The ordering of the returned vec corresponds to the ordering of `pages`.
    pub(crate) async fn get_rel_page_at_lsn_batched(
```

problame commented on pageserver/src/pgdatadir_mapping.rs:
```diff
@@ -191,26 +196,184 @@ impl Timeline {
+        // turn vectored_gets into a keyspace
+        let keyspace = {
+            // add_key reuqires monotonicity
+            vectored_gets.sort_unstable();
```
```suggestion
            // add_key requires monotonicity
```

VladLazar replied:
What ensures that the keys are sorted? I don't see anything in the batching logic to that effect.

VladLazar replied:
I rewrote this code to avoid an extra copy in https://github.com/neondatabase/neon/pull/9321/commits/a983a4907de45b23b4b26d879a6a6615913eadcb

problame commented on pageserver/src/pgdatadir_mapping.rs:
```diff
@@ -191,26 +196,184 @@ impl Timeline {
+        match self.get_vectored(keyspace, request_lsn, ctx).await {
+            Ok(results) => {
+                for (key, res) in results {
+                    if let Err(err) = &res {
+                        warn!(%key, ?err, "a key inside get_vectored failed with a per-key error");
+                    }
```
This could flood the logs

VladLazar replied:
https://github.com/neondatabase/neon/pull/9321/commits/0a554e0582f76b3d435229b48a6fc73c1fc884eb

problame commented on pageserver/src/pgdatadir_mapping.rs:
```diff
@@ -191,26 +196,184 @@ impl Timeline {
+                }
+            }
+            Err(err) => {
+                warn!(?err, "get_vectored failed with a global error, mapping that error to per-key failure");
```
Potential log flooder

VladLazar replied:
https://github.com/neondatabase/neon/pull/9321/commits/0a554e0582f76b3d435229b48a6fc73c1fc884eb

problame commented on pageserver/src/pgdatadir_mapping.rs:
```diff
@@ -191,26 +196,184 @@ impl Timeline {
+                    if let Err(err) = &res {
+                        warn!(%key, ?err, "a key inside get_vectored failed with a per-key error");
+                    }
+                    let mut interests = key_states.range_mut((key, 0)..(key.next(), 0)).peekable();
```
```suggestion
                    // We can technically have multiple interests for the same key.
                    // But get_vectored just returns one PageReconstructError per key.
                    // And PageReconstructError isn't Clone.
                    // => since this case is rare/nonexistent in practice, jump through
                    //    some hoops for better error messages returned to the client.
                    let mut interests = key_states.range_mut((key, 0)..(key.next(), 0)).peekable();
```

VladLazar replied:
Reworked this in https://github.com/neondatabase/neon/pull/9321/commits/a983a4907de45b23b4b26d879a6a6615913eadcb

problame commented on pageserver/src/pgdatadir_mapping.rs:
```diff
@@ -191,26 +196,184 @@ impl Timeline {
+                        GetVectoredError::Cancelled => {
+                            *state = KeyState::Done(Err(PageReconstructError::Cancelled));
+                        }
+                        // TODO: restructure get_vectored API to make this error per-key
+                        GetVectoredError::MissingKey(err) => {
```
This would be a nice preliminary or follow-up issue.

Do we have an epic for this work here? If not let's create one, link this PR and the other ones that commit our hackathon work, and list this TODO as a task item on that epic 🙏

VladLazar replied:
https://github.com/neondatabase/neon/issues/9382

problame commented on pageserver/src/pgdatadir_mapping.rs:
```diff
@@ -191,26 +196,184 @@ impl Timeline {
+                            *state =
+                                KeyState::Done(Err(anyhow::anyhow!("invalid LSN: {e:?}").into()));
+                        }
+                        // NB: this should never happen in practice because we limit MAX_GET_VECTORED_KEYS
+                        // TODO: we can prevent this error class by moving this check into the type system
```
Another task item for the epic

VladLazar replied:
https://github.com/neondatabase/neon/issues/9382

problame commented on pageserver/src/pgdatadir_mapping.rs:
```diff
@@ -191,26 +196,184 @@ impl Timeline {
 
-        let key = rel_block_to_key(tag, blknum);
-        version.get(self, key, ctx).await
+        // get the results into the order in which they were requested
+        let mut return_order: smallvec::SmallVec<[_; Timeline::MAX_GET_VECTORED_KEYS as usize]> =
+            smallvec::SmallVec::with_capacity(key_states.len());
+        return_order.extend(key_states.keys().map(|(key, idx)| (*key, *idx)));
+        return_order.sort_unstable_by_key(|(_, idx)| *idx);
+        let mut res = smallvec::SmallVec::with_capacity(key_states.len());
+        res.extend(return_order.into_iter().map(|key_states_key| {
```
There's one copy too many in here.

VladLazar replied:
Yeah. Reworked in https://github.com/neondatabase/neon/pull/9321/commits/a983a4907de45b23b4b26d879a6a6615913eadcb.

problame reviewed (commented):
The terminology is 50% debounce 50% batching.

I think `batching` is better, let's adjust the names.

As a person who doesn't mind wordy names, `page_service_server_side_batching` is how I'd call the functionality / code module.

Maybe `server_side_batch_timeout` for the config flag?

And `BatchedFeMessageS` instead of `DebouncedFeMessage` for the type?


---

During the hackathon, and now, I wonder(ed) whether we could refactor the batching loop into a Stream / $itertoolsadapter

Like, at the start of `handle_pagerequests`, split the `pgb` into reading and writing end (maybe fake that using `Arc<Mutex<Pgb>>`).
The wrap the reading end into the thing that implements Stream, spitting out `DebouncedFeMessage` as the strema item.

Orthogonal to that, can we please move the batching into a helper function so the high-level control flow of `handle_pagerequests` remains clear-ish?

---

I flagged a couple of places that warn! aggressively.
I actually don't know if we log getpage errors. If we do, then I think it's fine to just leave these `warn!` messages & dismiss the conversation.

VladLazar commented:
> The terminology is 50% debounce 50% batching.
> 
> I think `batching` is better, let's adjust the names.
> 
> As a person who doesn't mind wordy names, `page_service_server_side_batching` is how I'd call the functionality / code module.
> 
> Maybe `server_side_batch_timeout` for the config flag?
> 
> And `BatchedFeMessageS` instead of `DebouncedFeMessage` for the type?
> 

You're right. I'll touch up the names.

> During the hackathon, and now, I wonder(ed) whether we could refactor the batching loop into a Stream / $itertoolsadapter
> 
> Like, at the start of `handle_pagerequests`, split the `pgb` into reading and writing end (maybe fake that using `Arc<Mutex<Pgb>>`). The wrap the reading end into the thing that implements Stream, spitting out `DebouncedFeMessage` as the strema item.

I also wondered about this while going through the code, but resisted the temptation to rewrite.
Let's create an epic and add a follow-up in there.

> Orthogonal to that, can we please move the batching into a helper function so the high-level control flow of `handle_pagerequests` remains clear-ish?

Sure, I'll give it a go.

VladLazar commented on pageserver/src/page_service.rs:
```diff
@@ -591,124 +605,322 @@ impl PageServerHandler {
+                            &ctx,
+                        )
+                        // TODO: if we actually need to wait for lsn here, it delays the entire batch which doesn't need to wait
+                        .await
```
> Shouldn't make a difference while we leave debounce_timeout=None

That's my understanding as well.

VladLazar commented on pageserver/src/page_service.rs:
```diff
@@ -591,124 +605,322 @@ impl PageServerHandler {
+                    (Some(debounce_timeout), Some(started_at)) => futures::future::Either::Left(
+                        tokio::time::sleep_until((started_at + debounce_timeout).into()),
+                    ),
+                    _ => futures::future::Either::Right(futures::future::pending()),
```
> Did you re-investigate this or are you ok with the crutch?

Nope. Okay with leaving as is for now. Same pattern is used in other places.

VladLazar commented on pageserver/src/page_service.rs:
```diff
@@ -591,124 +605,322 @@ impl PageServerHandler {
+                    (Some(_), this_msg) => {
+                        // by default, don't continue batching
+                        break Some(this_msg);
+                    }
```
Histogram for batch sizes? Will punt for now though.

VladLazar replied:
https://github.com/neondatabase/neon/issues/9380

VladLazar commented:
* Switched naming to batching
* Looked into logging. We already log on both pagestream and ingest flows, so I've removed the redundant logging.
* Reworked response filling to work around one copy
* Refactored batching into a separate method.

problame commented on pageserver/src/page_service.rs:
```diff
@@ -591,133 +842,170 @@ impl PageServerHandler {
+                        pages,
+                    } => {
+                        fail::fail_point!("ps::handle-pagerequest-message::getpage");
+                        // shard_id is filled in by the handler
```
stray comment 
```suggestion
```

problame commented on pageserver/src/page_service.rs:
```diff
@@ -963,60 +1251,31 @@ impl PageServerHandler {
 
-        Ok(PagestreamBeMessage::GetPage(PagestreamGetPageResponse {
-            page,
+        // TODO: This copies the buffers
+        Vec::from_iter(pages.into_iter().map(|page| {
```
This TODO is obsolete, isn't it?


```suggestion
        Vec::from_iter(pages.into_iter().map(|page| {
```

problame replied:
294e1890f7380e3d60f4c2989a1efebf6544bc15

problame commented on pageserver/src/pgdatadir_mapping.rs:
```diff
@@ -191,26 +196,197 @@ impl Timeline {
+    pub(crate) async fn get_rel_page_at_lsn_batched(
+        &self,
+        pages: smallvec::SmallVec<[(RelTag, BlockNumber); 1]>,
+        version: Version<'_>,
+        ctx: &RequestContext,
+    ) -> Vec<Result<Bytes, PageReconstructError>> {
+        debug_assert_current_span_has_tenant_and_timeline_id();
+        let request_lsn = match version {
+            Version::Lsn(lsn) => lsn,
+            Version::Modified(_) => panic!("unsupported"),
+        };
```
Let's remove `Version` from the signature and take an `Lsn` instead.
That eliminates this panic.

problame commented on pageserver/src/pgdatadir_mapping.rs:
```diff
@@ -191,26 +196,197 @@ impl Timeline {
+        pages: smallvec::SmallVec<[(RelTag, BlockNumber); 1]>,
+        version: Version<'_>,
+        ctx: &RequestContext,
+    ) -> Vec<Result<Bytes, PageReconstructError>> {
```
Should return a smallovec<1> here as well, otherwise this PR is a change in allocation behavior.

problame replied:
4dba135fe8e2cdac04dec1d9d2568769a24d2715

problame commented on pageserver/src/pgdatadir_mapping.rs:
```diff
@@ -191,26 +196,197 @@ impl Timeline {
+        }
+
+        let keyspace = {
+            // add_key reuqires monotonicity
```
```suggestion
            // add_key requires monotonicity
```

problame commented on pageserver/src/pgdatadir_mapping.rs:
```diff
@@ -191,26 +196,197 @@ impl Timeline {
+                }
+            }
+            Err(err) => {
+                // this cannot really happen because get_vectored only errors globally on invalid LSN or too large batch size
```
```suggestion
                // this cannot really happen because get_vectored only errors globally on invalid LSN or too large batch size
                // (We enforce the max batch size outside of this function, in the code that constructs the batch request.)
```

problame reviewed (changes requested):
I see a lot of plumbing-through of the `server_side_batch_timeout` variable.
Personally I wouldn't mind just plumbing through the `&'static PageServerConf`.
Or building a `PageServiceConf` object out of PageServerConf.
Easily punt-able though.

---

I think the only two things I'm hard-requesting-changes for is the use of smallvec<1> for the return value of `read_batch_from_connection`, and the elimination of the `panic!(unsupported)`

problame commented on pageserver/src/page_service.rs:
```diff
@@ -516,10 +531,31 @@ impl From<WaitLsnError> for QueryError {
+
+enum BatchOrEof {
+    // TODO: Use smallvec. May have at most two entries
+    Batch(Vec<BatchedFeMessage>),
```
39510b24a1db0c27bc3a8a372690843f174a109b
